### ISSUE TITLE
[std] file size should be u64 instead of usize

### ISSUE THREAD

**Issue by thejoshwolfe** [ziglang/zig#637]
tldr: Don't conflate virtual address space size and file size.

Currently, `std.io.File.getEndPos` and `std.io.File.seekTo` etc. use `usize` for the file size and file offset integer type. The underlying Posix api uses a vaguely defined `off_t` type, which is only guaranteed to be signed. Unless we want to bubble this vagueness through the zig std library, we're going to have potentially mismatched types for this API, which means the std library needs to do some conversion in both directions.

Using `usize` means that it's impossible to get the file size of files larger than 4GB on 32bit platforms. In fact, it is a language-level assertion failure to even attempt it, which seems like a bug. It's possible that files that large are indeed impossible on a particular system, but that assumption should not be in the zig std library api.

Proposal: Make file sizes and offsets be `u64` instead of `usize`. When converting in either direction, there's a potential for out of bounds values. This is at least in part due to `off_t` being signed, and we usually don't want to use signed values for this information. Do an explicit bounds check, and return an error for out of bounds values.

I propose `u64` instead of some other size, because it's big enough for any reasonable file in the modern world.

One reason for wanting a fixed size instead of a comptime dynamic size is that i want to be able to format the size as a string like this: `format("{x16}", offset)`. So if we wanted a dynamic size, we'd also want some way of making the formatting dynamic, like exposing string literals for the size in bytes of various things so you can `format("{x" ++ std.os.size_of_file_size_str ++ "}", offset)` (only then to get parsed at comptime by `format()`, ugh.).

**Comment by andrewrk** [comment]
> In fact, it is a language-level assertion failure to even attempt it, which seems like a bug. It's possible that files that large are indeed impossible on a particular system, but that assumption should not be in the zig std library api.

This part isn't true. If we have to downcast a bigger integer in file size detection code, we return `error.FileTooBig` or equivalent, as seen here:
https://github.com/ziglang/zig/blob/e5956f23ca702b79a3a4b0f0440a2fe88e0231e5/std/os/file.zig#L280

**Comment by LemonBoy** [comment]
I wholeheartedly agree and I'd go even further and propose the use of 64-bit syscalls (eg. fstat64, llseek) on 32 bit systems.

**Comment by andrewrk** [comment]
I agree as well. This is planned and patches welcome.
